core/arch/arm/kernel/elf_load.c

```diff
@@ -0,0 +1,371 @@
+		return TEE_ERROR_BAD_STATE;
+	if (offs == state->next_offs)
+		return TEE_SUCCESS;
+
+	if (offs > state->nwdata_len)
+		return TEE_ERROR_SECURITY;
+
+	res = crypto_ops.hash.update(state->hash_ctx, state->hash_algo,
+			state->nwdata + state->next_offs,
+			offs - state->next_offs);
+	if (res != TEE_SUCCESS)
+		return res;
+	state->next_offs = offs;
+	return res;
+}
+
+static TEE_Result copy_to(struct elf_load_state *state,
+			void *dst, size_t dst_size, size_t dst_offs,
+			size_t offs, size_t len)
+{
+	TEE_Result res;
+
+	res = advance_to(state, offs);
+	if (res != TEE_SUCCESS)
+		return res;
+	if (!len)
+		return TEE_SUCCESS;
+
+	if (len > dst_size || (len + dst_offs) > dst_size)
+		return TEE_ERROR_SECURITY;
+
+	if (!core_is_buffer_inside(state->nwdata + offs, len,
+				   state->nwdata, state->nwdata_len))
+		return TEE_ERROR_SECURITY;
+
+	memcpy((uint8_t *)dst + dst_offs, state->nwdata + offs, len);
+	res = crypto_ops.hash.update(state->hash_ctx, state->hash_algo,
+				      (uint8_t *)dst + dst_offs, len);
+	if (res != TEE_SUCCESS)
+		return res;
+	state->next_offs = offs + len;
+	return res;
+}
+
+static TEE_Result alloc_and_copy_to(void **p, struct elf_load_state *state,
+			size_t offs, size_t len)
+{
+	TEE_Result res;
+	void *buf;
+
+	buf = malloc(len);
+	if (!buf)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	res = copy_to(state, buf, len, 0, offs, len);
+	if (res == TEE_SUCCESS)
+		*p = buf;
+	else
+		free(buf);
+	return res;
+}
+
+TEE_Result elf_load_init(void *hash_ctx, uint32_t hash_algo, uint8_t *nwdata,
+			size_t nwdata_len, struct elf_load_state **ret_state)
+{
+	struct elf_load_state *state;
+
+	state = calloc(1, sizeof(*state));
+	if (!state)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	state->hash_ctx = hash_ctx;
+	state->hash_algo = hash_algo;
+	state->nwdata = nwdata;
+	state->nwdata_len = nwdata_len;
+	*ret_state = state;
+	return TEE_SUCCESS;
+}
+
+TEE_Result elf_load_head(struct elf_load_state *state, size_t head_size,
+			void **head, size_t *vasize)
+{
+	TEE_Result res;
+	size_t n;
+	void *p;
+	Elf32_Ehdr *ehdr;
+
+	/*
+	 * The ELF resides in shared memory, to avoid attacks based on
+	 * modifying the ELF while we're parsing it here we only read each
+	 * byte from the ELF once. We're also hashing the ELF while reading
+	 * so we're limited to only read the ELF sequentially from start to
+	 * end.
+	 */
+
+	res = alloc_and_copy_to(&p, state, 0, sizeof(Elf32_Ehdr));
+	if (res != TEE_SUCCESS)
+		return res;
+	ehdr = p;
+	state->ehdr = ehdr;
+
+	if (!IS_ELF(*ehdr) ||
+	    ehdr->e_ident[EI_VERSION] != EV_CURRENT ||
+	    ehdr->e_ident[EI_CLASS] != ELFCLASS32 ||
+	    ehdr->e_ident[EI_DATA] != ELFDATA2LSB ||
+	    ehdr->e_ident[EI_OSABI] != ELFOSABI_NONE ||
+	    ehdr->e_type != ET_DYN || ehdr->e_machine != EM_ARM ||
+	    ehdr->e_phentsize != sizeof(Elf32_Phdr) ||
+	    ehdr->e_shentsize != sizeof(Elf32_Shdr))
+		return TEE_ERROR_BAD_FORMAT;
+
+	/*
+	 * Program headers are supposed to be arranged as:
+	 * LOAD [0] : .ta_head ...
```

Comment:
Nit-picking: it may be better to use the full name of the ELF constants in the comments, that is: `PT_LOAD`, `PT_DYNAMIC`.